src/main/kotlin/io/github/nomisrev/routes/profile.kt

```diff
@@ -0,0 +1,49 @@
+package io.github.nomisrev.routes
+
+import arrow.core.Either
+import arrow.core.raise.either
+import io.github.nomisrev.IncorrectJson
+import io.github.nomisrev.service.ProfileService
+import io.ktor.http.HttpStatusCode
+import io.ktor.resources.Resource
+import io.ktor.server.application.ApplicationCall
+import io.ktor.server.application.call
+import io.ktor.server.resources.get
+import io.ktor.server.routing.Route
+import io.ktor.server.util.getOrFail
+import io.ktor.util.pipeline.PipelineContext
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.MissingFieldException
+import kotlinx.serialization.Serializable
+
+@JvmInline @Serializable value class GetProfile(val username: String)
+
+@Resource("/profiles")
+data class ProfileResource(val parent: RootResource = RootResource) {
+  @Resource("/{$USERNAME}")
+  data class Username(val parent: ProfileResource = ProfileResource(), val username: String)
+}
+
+fun Route.profileRoutes(
+  profileService: ProfileService,
+) {
+  get<ProfileResource.Username> {
+    either {
+        val username = parameter(USERNAME, ::GetProfile).bind().username
+        profileService.getProfile(username).bind()
+      }
+      .respond(HttpStatusCode.OK)
+  }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+private inline fun <reified A : Any> PipelineContext<Unit, ApplicationCall>.parameter(
+  parameter: String,
+  noinline toRight: (String) -> A,
+): Either<IncorrectJson, A> =
+  Either.catchOrThrow<MissingFieldException, A> {
+      call.parameters.getOrFail(parameter).let { toRight(it) }
+    }
+    .mapLeft(::IncorrectJson)
```

Comment:
This is not actually `IncorrectJson` 🤔 `getOrFail` throws `MissingRequestParameterException` and not `MissingFieldException`.

So I think we should replace this as:

```kotlin
call.parameters.get(parameter)?.let(::toRight)?.right() ?: MissingParameter(name)
```

And then we'll have to introduce `MissingPathParameter` in `DomainError.kt`. I think we can do:

```kotlin
data class MissingParameter(val name: String): ValidationError
```

We then need to resolve it as `unprocessable` inside `error.kt` by adding the code below in `respond`.
```kotlin
is MissingParameter -> unprocessable("Missing ${error.name} parameter in request")
```

Reply:
Addressed in [d952243](https://github.com/nomisRev/ktor-arrow-example/pull/186/commits/d952243ebbb77bbb2d293ef51b95f0a7cfab8dc3)